### ISSUE TITLE
mesa not building for x86_64-w64-mingw32.shared

### ISSUE THREAD

**Issue by LubosD** [mxe/mxe#2561]
I'm trying to build whole qt5 for Windows x86-64:

```
make MXE_TARGETS="x86_64-w64-mingw32.shared" qt5
```

The build fails, however:

```
[build]       mesa                   x86_64-w64-mingw32.shared

Failed to build package mesa for target x86_64-w64-mingw32.shared!
------------------------------------------------------------
    from mako.template import Template
ImportError: No module named mako.template
scons: *** [build/windows-x86_64/mesa/main/format_fallback.c] Error 1
scons: building terminated because of errors.
make[1]: *** [Makefile:846: build-only-mesa_x86_64-w64-mingw32.shared] Error 2
make[1]: Leaving directory '/mxe'
```

Whole build log: https://pastebin.com/XMVH4H2y

**Comment by jonaski** [comment]
Did you try to install python mako?

**Comment by LubosD** [comment]
@jonaski Nope, I just followed the [Requirements for Debian](https://mxe.cc/#requirements-debian). I'll do that and report back.

**Comment by LubosD** [comment]
Adding `python-mako` to the package list in my Dockerfile helped. So this package should be added in the requirements list.

**Comment by cgilles** [comment]
Hi,

I can reproduce this problem under Mageia 6 to build whole digiKam installer with MXE.

I installed python2 and python3 mako packages and this do not solve this issue:

```
[build]       libmysqlclient         i686-w64-mingw32.shared
[done]        libmysqlclient         i686-w64-mingw32.shared                                  161496 KiB     1m47.734s
[download]    mesa-18.3.6.tar.xz
[download]    scons-local-3.1.2.tar.gz
[build]       mesa                   i686-w64-mingw32.shared

Failed to build package mesa for target i686-w64-mingw32.shared!
------------------------------------------------------------
cc1: sorry, unimplemented: 64-bit mode not compiled in
scons: *** [build/windows-x86_64/gallium/state_trackers/wgl/stw_ext_context.o] Error 1
scons: building terminated because of errors.
Makefile:827: recipe for target 'build-only-mesa_i686-w64-mingw32.shared' failed
make[1]: *** [build-only-mesa_i686-w64-mingw32.shared] Error 2
make[1]: Leaving directory '/mnt/data/GIT/7.x/project/bundles/mxe/build.win32'
real    0m33.908s
user    0m32.470s
sys     0m1.817s
------------------------------------------------------------
[log]      /mnt/data/GIT/7.x/project/bundles/mxe/build.win32/log/mesa_i686-w64-mingw32.shared

Makefile:827: recipe for target '/mnt/data/GIT/7.x/project/bundles/mxe/build.win32/usr/i686-w64-mingw32.shared/installed/mesa' failed
make: *** [/mnt/data/GIT/7.x/project/bundles/mxe/build.win32/usr/i686-w64-mingw32.shared/installed/mesa] Error 1
FAILED COMMAND: make MXE_TARGETS=$MXE_BUILD_TARGETS cmake gettext freeglut libxml2 libxslt libjpeg-turbo libpng tiff boost expat lcms liblqr-1 eigen jasper zlib mman-win32 pthreads qtbase qttranslations qtimageformats qtwebkit qttools qtwinextras qtscript x265 ffmpeg openal libical imagemagick
FAILED COMMAND: ./01-build-mxe.sh
[gilles@localhost mxe (master)]$ 
```

**Comment by nschlia** [comment]
> I installed python2 and python3 mako packages and this do not solve this issue:

Same for me, "apt-get update && apt-get dist-upgrade && apt-get install python-mako", no change. Build still fails with the above error.

Debian 10 - Debian 4.19.132-1 (2020-07-24)

**Comment by hyOzd** [comment]
I think this issue is created by recent commits. Because I can build with this commit: 3735aa8020609e680ca6867e3f4835d8a06efa9b .

I'm building on Ubuntu Bionic (Docker). I've also tried installing `python-mako` package. It didn't help. BTW I don't have to install it when building 3735aa802

**Comment by mabrand** [comment]
@mardy : Could you please take a  look at this?

**Comment by mardy** [comment]
> @mardy : Could you please take a look at this?

I'll try to find some time for it, but if it's urgent I'll certainly not complain if someone else has a look at it meanwhile :-)

**Comment by mabrand** [comment]
x86_64-w64-mingw32.static and x86_64-w64-mingw32.shared build without error for me, but I can reproduce the problem for i686-w64-mingw32.shared.

**Comment by mabrand** [comment]
Building mesa for i686-w64-mingw32 should work now. Do we have a volunteer to add the requirement python-mako to docs/index.html?

**Comment by cgilles** [comment]
@mabrand :I will restart whole digiKam mxe build to check if problem is solved

**Comment by cgilles** [comment]
@mabrand : I confirm that a fresh MXE digiKam build fix the problem. Python-mako package must be installed first.

Gilles Caulier

**Comment by nschlia** [comment]
Builds for me now, seems fixed. THX!

**Comment by cgilles** [comment]
yes, as nschlia said, a BIG thanks to maintain to jewel named MXE...

Remember that whole digiKam installers 32 and 64 bits are all cross compiled under Linux. This take around 4 hours on my host computer (i7 - 8 cores - 32Gb). This cross compile all, including Qt5, KDE frameworks, OpenCV, boost, NSIS, and whole digiKam (1.5M lines of C++ code). 

And this will stay like this for a while, as MSVC + Windows is just a pain for developers.

Best regards

Gilles Caulier

**Comment by scorpio810** [comment]
Well, but Debian removed in unstable python 2 packages, not working with python3-mako for me ...

**Comment by tonytheodore** [comment]
Scons uses the same python version it was invoked with, and it currently runs with python2 so python3-mako won't be recognised.

I'll have a look at updating scons-local and removing any python2 references.

**Comment by mmorais-idealsumario** [comment]
I was trying to make qt5 for x86_64-w64-mingw32.static with the same issue. Using opensuse tumbleweed which no longer has mako for python2, so basically I had to install it with pip2 for it to work.

**Comment by tonytheodore** [comment]
We'll bundle mako (and its req MarkupSafe) - there's too many variants to add to native requirements.

**Comment by aquiles2k** [comment]
What package do i have to install with brew on mac for this to work? Python?

**Comment by jonaski** [comment]
I worked around it by patching scons to point to python3
https://raw.githubusercontent.com/strawberrymusicplayer/strawberry-mxe/master/src/scons-local-1-fixes.patch

**Comment by jonaski** [comment]
You need python3-mako, not sure about brew

**Comment by scorpio810** [comment]
Thanks Jonaski I try it maybe this week. ;-)

**Comment by Scorbie** [comment]
N.B. While this is being fixed a workaround would be just installing mako on python2:
1. Install pip as per: https://pip.pypa.io/en/stable/installing/#installing-with-get-pip-py
2. Run `python2 -m pip install mako --user`

**Comment by aquiles2k** [comment]
@tonytheodore Any hints on what to do when crosscompiling on mac?

**Comment by mabrand** [comment]
I ran into the same problem on Ubuntu 20.10 with python3-mako installed. Following jonaski tips, I updated scons and mesa and made adjustments to use python 3. This works for me. Thank you jonaski.

All the changes are here:
https://github.com/mxe/mxe/commit/92634116e8013bfdae3af03f13fa9b77231ea3e0

This isn't general enough to work on all systems, but I think it will help on "python 3" systems affected by the problem.
